lib/scholar/neighbors/graph.ex

```diff
@@ -0,0 +1,182 @@
+defmodule Scholar.Neighbors.Graph do
+  import Nx.Defn
+  require Nx
+  alias Scholar.Neighbors.RandomProjectionForest, as: Forest
+
+  knn_schema = [
+    k: [
+      required: true,
+      type: :pos_integer,
+      doc: "The number of neighbors in the graph."
+    ]
+  ]
+
+  # TODO: Rename this
+  approximate_knn_schema =
+    knn_schema ++
+      [
+        min_leaf_size: [
+          type: :pos_integer,
+          doc: "The minimum number of points in ..."
+        ],
+        num_trees: [
+          type: :pos_integer,
+          doc: "The number of trees in random projection forest."
+        ],
+        num_iters: [
+          type: :non_neg_integer,
+          default: 1,
+          doc: "The number of times to perform neighborhood expansion."
+        ],
+        key: [
+          type: {:custom, Scholar.Options, :key, []},
+          doc: """
+          Used for random number generation in parameter initialization.
+          If the key is not provided, it is set to `Nx.Random.key(System.system_time())`.
+          """
+        ]
+      ]
+
+  @knn_schema NimbleOptions.new!(knn_schema)
+  @approximate_knn_schema NimbleOptions.new!(approximate_knn_schema)
+
+  deftransform exact_brute_force(x, opts) do
+    opts = NimbleOptions.validate!(opts, @knn_schema)
+    exact_brute_force_n(x, opts)
+  end
+
+  defn exact_brute_force_n(x, opts) do
+    k = opts[:k]
+    # TODO: Add other metrics other than Euclidean
+    x
+    |> Nx.new_axis(1)
+    |> Nx.subtract(Nx.new_axis(x, 0))
+    |> Nx.pow(2)
+    |> Nx.sum(axes: [2])
```

Comment:
Here you have all other metrics implemented
```elixir
  defn pairwise_dist(x, opts) do
    {num_samples, num_features} = Nx.shape(x)
    broadcast_shape = {num_samples, num_samples, num_features}


    t1 =
      x
      |> Nx.reshape({1, num_samples, num_features})
      |> Nx.broadcast(broadcast_shape)


    t2 =
      x
      |> Nx.reshape({num_samples, 1, num_features})
      |> Nx.broadcast(broadcast_shape)


    case opts[:metric] do
      :squared_euclidean ->
        Distance.pairwise_squared_euclidean(x)


      :euclidean ->
        Distance.pairwise_euclidean(x)


      :manhattan ->
        Distance.manhattan(t1, t2, axes: [2])


      :cosine ->
        Distance.pairwise_cosine(x)


      :chebyshev ->
        Distance.chebyshev(t1, t2, axes: [2])
    end
  end
```

Reply:
Thanks! I've seen this somewhere implemented somewhere else, perhaps we can move it in a shared module?

Reply:
Yeah, this function is directly used in `t-SNE`. We may consider having this as a separate function.